fix(perf): Improve sidebar accordion active state II

gggritso opened the pull request:
Follow-up to #56762. Since that PR, the `SidebarItem` elements got a `Feature` wrapper, so we lost the functionality. In this PR, I make sure to actually find the right elements to check again.

Recursion is a bit overkill here, but to be honest the best solution would probably be a totally different, more explicit logic which is a much bigger refactor.

This component is a little tricky to unit test, so I'm open to suggestions. Maybe test the helper function at least?

gggritso committed:
Find `SidebarItem` recursively

When determining sidebar link active state, check the children
recursively until we find `SidebarItem`, which might not be the
top-level child.

billyvg commented on static/app/components/sidebar/sidebarAccordion.tsx:
```diff
@@ -78,6 +81,65 @@ function SidebarAccordion({children, ...itemProps}: SidebarAccordionProps) {
+  return found;
+}
+
+function isAReactElement(element: React.ReactNode): element is React.ReactElement {
```
Can you use https://react.dev/reference/react/isValidElement?

gggritso replied:
Honestly embarrassing I didn't see this function being used _in this file_. Yes! Thank you! Changed

gggritso committed:
Use `isValidElement` instead

It's literally the same thing but better.

JonasBa reviewed (commented):
Do you know if anyone in the codebase is possibly doing something similar? It might just be worth placing this inside a util file and slightly modifying the API if so. It wouldn't surprise me if we are doing something similar :D

gggritso committed:
Simplify type check

Co-authored-by: Jonas <jonas.badalic@sentry.io>

gggritso commented on static/app/components/sidebar/sidebarAccordion.tsx:
```diff
@@ -78,6 +81,37 @@ function SidebarAccordion({children, ...itemProps}: SidebarAccordionProps) {
 
 export {SidebarAccordion};
 
+function recursivelyFindChildren(
```
Fair! Went with `findChildElementsInTree` which I think is even clearer 👍🏻

JonasBa replied:
Yaasssss, that is even better indeed!

gggritso committed:
Clarify function name

gggritso committed:
Improve conditional logic

1. Clean up component name extraction a bit
2. Return early if the child is found

gggritso commented:
@JonasBa thanks for the feedback! I took a look around the codebase and there's a handful of `React.Children` usage, but they all look like clean iterations or simple checks. I couldn't find any recursive searches. I'm happy to extract this util if you think it's useful, though.

JonasBa commented:
> @JonasBa thanks for the feedback! I took a look around the codebase and there's a handful of `React.Children` usage, but they all look like clean iterations or simple checks. I couldn't find any recursive searches. I'm happy to extract this util if you think it's useful, though.

Nah, lets not overthink it. 🚢

JonasBa commented on static/app/components/sidebar/sidebarAccordion.tsx:
```diff
@@ -78,6 +81,37 @@ function SidebarAccordion({children, ...itemProps}: SidebarAccordionProps) {
 
 export {SidebarAccordion};
 
+function recursivelyFindChildren(
```
I was thinking was that if we had a slot like API here, we would be able to avoid searching the tree for children (as they would self register with the parent). It wouldn't be slots in the traditional sense that they'd have to be unique, but they could just register with the parent component

gggritso committed:
Rename vague variable

gggritso committed:
Mark active sidebar item with `aria-current`

gggritso committed:
Add spec

gggritso commented:
@JonasBa cc @vuluongj20 I ended up adding the `aria-current` attribute to the active sidebar item in this PR, so I could write a functional test.

The weird thing is that the label is currently set to the `children` so for accordions, it means the child names are part of the label. I opted not to fix that in this PR since I don't know the details
